src/context.rs

```diff
@@ -0,0 +1,230 @@
+use futures::channel::oneshot;
+use futures::future::{pending, select, BoxFuture, FutureExt, Shared};
+use tokio::sync::mpsc;
+use tokio::time::{delay_for, Duration};
+
+pub struct CancelHandle {
+    cancel_fn: Box<dyn FnOnce()>,
+}
```

Comment:
It looks like you are using this to be able to cancel interest in a particular future being completed. Have you looked at the [`abortable`](https://docs.rs/futures/0.3.5/futures/future/fn.abortable.html) future wrapper in futures? It definitely doesn't function quite the same way this `CancelHandle` does, but I think it would accomplish the same goal.

Reply:
I saw this API, and I was planning to use it for "brutal termination" (e.g. `kill -9`) functionality. It doesn't seem you get a signal on the aborted future to react appropietly.

The goal of this this API is to signal a "we are getting kicked out" on the Future task, and we need to do cleanup now. There are many ways to accomplish this, I decided to replicate golang's approach, but it doesn't need to be like that.

Maybe we can come back to this (remove it if necessary) once we exercise how this API is used?

Reply:
Coming back to it for sure makes sense, I just wanted to make sure you were aware of it and had made an intentional choice.

Wanting to get a signal for graceful cleanup definitely makes sense, but I don't really see how this API actually accomplishes that for notifying the inner future (right now). Since the context can hold arbitrary futures I think a new trait would need to be introduced with a `poll_cancel` method or similar if the future is allowed to take some async action to perform cleanup. But as you say, I think that can wait until this API is actually getting used by other code.

Reply:
Nevermind, I was able to use abortable